common/matrix/csr_kernels.hpp.inc

```diff
@@ -494,95 +494,95 @@ __global__ __launch_bounds__(spmv_block_size) void abstract_classical_spmv(
 }
 
 
-template <typename IndexType>
+template <int subwarp_size, typename IndexType>
 __global__ __launch_bounds__(default_block_size) void spgeam_nnz(
     const IndexType *a_row_ptrs, const IndexType *a_col_idxs,
     const IndexType *b_row_ptrs, const IndexType *b_col_idxs,
     IndexType num_rows, IndexType *nnz)
 {
-    constexpr auto sentinel = device_numeric_limits<IndexType>::max;
-    auto row = threadIdx.x + blockDim.x * static_cast<size_type>(blockIdx.x);
+    auto row =
+        (threadIdx.x + blockDim.x * size_type(blockIdx.x)) / subwarp_size;
```

Comment:
nit: can you keep the `static_cast`?
Alternatively, if `blockIdx.x` is actually an unsigned, this should also work:
```suggestion
        (threadIdx.x + blockDim.x * size_type{blockIdx.x}) / subwarp_size;
```

Reply:
I guess that we've had this discussion often enough that it would make sense to find either a guideline or a non-guideline (i.e., we don't care) for this type of operation. To me, this is heavily connected with https://github.com/ginkgo-project/ginkgo/issues/430 since ideally, I would like to 1. maybe refactor our kernels so these types of calculations are executed in a device function or use cooperative_group grid groups and 2. get rid of as many static_cast (and of course functional-style cast) operations by making our size types consistent everywhere. I will elaborate a bit in #430.

After thinking about it for a while, I slightly agree with your viewpoint, but also to find the amount of static_casts in our code base too big.